infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/RuntimeHangingDetector.java

```diff
@@ -0,0 +1,212 @@
+   */
+  public synchronized void trackStarting(KubernetesInternalRuntime runtime, long timeoutMin) {
+    String workspaceId = runtime.getContext().getIdentity().getWorkspaceId();
+    WaitStatusChangedTask waitStartingChangedTask =
+        new WaitStatusChangedTask(
+            runtime, WorkspaceStatus.STARTING, this::handleHangingStartingRuntime);
+
+    LOG.debug(
+        "Registered a task to check runtime '{}' to become RUNNING OR STOPPED after {} minutes",
+        workspaceId,
+        timeoutMin);
+    workspaceId2Task.put(workspaceId, waitStartingChangedTask);
+    getTimer().schedule(waitStartingChangedTask, TimeUnit.MINUTES.toMillis(timeoutMin));
+  }
+
+  private void handleHangingStartingRuntime(KubernetesInternalRuntime runtime) {
+    RuntimeIdentity runtimeId = runtime.getContext().getIdentity();
+    try {
+      eventPublisher.sendAbnormalStoppingEvent(
+          runtimeId,
+          WorkspaceStatus.STARTING,
+          "Workspace is not started in time. Trying interrupt runtime start");
+      runtime.stop(emptyMap());
+      eventPublisher.sendAbnormalStoppedEvent(runtimeId, "Workspace start reached timeout");
+      LOG.info(
+          "Start of hanging runtime '{}:{}:{}' is interrupted",
+          runtimeId.getWorkspaceId(),
+          runtimeId.getEnvName(),
+          runtimeId.getOwnerId());
+    } catch (InfrastructureException e) {
+      LOG.error(
+          "Error occurred during start interruption of hanging runtime '{}:{}:{}'. Error: {}",
+          runtimeId.getWorkspaceId(),
+          runtimeId.getEnvName(),
+          runtimeId.getOwnerId(),
+          e.getMessage(),
+          e);
+    }
+  }
+
+  /**
+   * Should be invoked after recovering of {@link WorkspaceStatus#STOPPING} runtime to track it and
+   * avoid hanging.
+   *
+   * @param runtime runtime to track
+   * @param timeoutMin timeout before which runtime should change its state
+   */
+  public synchronized void trackStopping(KubernetesInternalRuntime runtime, long timeoutMin) {
+    String workspaceId = runtime.getContext().getIdentity().getWorkspaceId();
+    WaitStatusChangedTask waitStoppingChangedTask =
+        new WaitStatusChangedTask(
+            runtime, WorkspaceStatus.STOPPING, this::handleHangingStoppingRuntime);
+
+    LOG.debug(
+        "Registered a task to check workspace {} to become STOPPED after {} minutes",
+        workspaceId,
+        timeoutMin);
+    workspaceId2Task.put(workspaceId, waitStoppingChangedTask);
+    getTimer().schedule(waitStoppingChangedTask, TimeUnit.MINUTES.toMillis(timeoutMin));
+  }
+
+  private void handleHangingStoppingRuntime(KubernetesInternalRuntime runtime) {
+    RuntimeIdentity runtimeId = runtime.getContext().getIdentity();
+    try {
+      eventPublisher.sendAbnormalStoppingEvent(
+          runtimeId,
+          WorkspaceStatus.STOPPING,
+          "Workspace is not stopped in time. Trying to stop it forcibly");
+      runtime.internalStop(emptyMap());
```

Comment:
In this class, we use a single thread for ALL the workspaces in the system that needs to be tracked. Calling `internalStop` here would execute all the cleanup operations in that thread. I'm concerned on how this may influence the system in a case we get request timeout on calls to k8s API or even waiting for deployments removal that happens in that thread too would be too long.

Reply:
Good catch! I'll rework it to submit stop task to `WorkspaceSharedPool`